include/trieste/intrusive_ptr.h

```diff
@@ -0,0 +1,350 @@
+    // 0, as if the object was constructed from scratch, so different
+    // intrusive_ptr can take ownership of the new object.
+    struct copyable_refcount final
+    {
+    private:
+      std::atomic<size_t> value = 0;
+
+    public:
+      constexpr copyable_refcount(size_t value_) : value{value_} {}
+
+      constexpr copyable_refcount() = default;
+      constexpr copyable_refcount(const copyable_refcount&) : value{0} {}
+      constexpr copyable_refcount(copyable_refcount&&) : value{0} {}
+
+      operator size_t() const
+      {
+        return value;
+      }
+
+      copyable_refcount& operator+=(size_t inc)
+      {
+        value += inc;
+        return *this;
+      }
+
+      size_t fetch_sub(size_t dec)
+      {
+        return value.fetch_sub(dec);
+      }
+    };
+  }
+
+  template<typename T>
+  struct intrusive_refcounted_traits
+  {
+    static constexpr void intrusive_inc_ref(T* ptr)
+    {
+      ptr->intrusive_inc_ref();
+    }
+
+    static constexpr void intrusive_dec_ref(T* ptr)
+    {
+      ptr->intrusive_dec_ref();
+    }
+  };
+
+  template<typename T>
+  struct intrusive_ptr final
+  {
+  private:
+    T* ptr;
+
+    constexpr void inc_ref()
+    {
+      if (ptr)
+      {
+        intrusive_refcounted_traits<T>::intrusive_inc_ref(ptr);
+      }
+    }
+
+    constexpr void dec_ref()
+    {
+      if (ptr)
+      {
+        intrusive_refcounted_traits<T>::intrusive_dec_ref(ptr);
+        ptr = nullptr;
+      }
+    }
+
+  public:
+    template<typename... Args>
+    static intrusive_ptr<T> make(Args&&... args)
+    {
+      return intrusive_ptr(new T(std::forward<Args>(args)...));
+    }
+
+    constexpr intrusive_ptr() : ptr{nullptr} {}
+
+    constexpr intrusive_ptr(std::nullptr_t) : ptr{nullptr} {}
+
+    constexpr explicit intrusive_ptr(T* ptr_) : ptr{ptr_}
+    {
+      inc_ref();
+    }
+
+    template<typename U>
+    constexpr intrusive_ptr(const intrusive_ptr<U>& other) : ptr{other.ptr}
+    {
+      inc_ref();
+    }
+
+    template<typename U>
+    constexpr intrusive_ptr(intrusive_ptr<U>&& other) : ptr{other.release()}
+    {}
+
+    constexpr intrusive_ptr(const intrusive_ptr<T>& other) : ptr{other.ptr}
+    {
+      inc_ref();
+    }
+
+    constexpr intrusive_ptr(intrusive_ptr<T>&& other) : ptr{other.release()} {}
+
+    constexpr intrusive_ptr<T>& operator=(const intrusive_ptr<T>& other)
+    {
+      // Self-assignment case, don't bother touching refcounts then
+      if (ptr == other.ptr)
+      {
+        return *this;
+      }
+
+      intrusive_ptr<T> tmp;
+      // Don't actually inc_ref, but putting old ptr in tmp lets us leverage the
+      // built in dec_ref with null checks below.
+      tmp.ptr = ptr;
+      ptr = other.ptr;
+      inc_ref();
+      // tmp gets dec_ref here, potentially destroying the value at old ptr
+      return *this;
+    }
+
+    constexpr intrusive_ptr<T>& operator=(intrusive_ptr<T>&& other)
+    {
+      std::swap(ptr, other.ptr);
```

Comment:
Is this really a swap? Should we be putting `ptr` into `other`?

```suggestion
intrusive_ptr<T> old{ptr};  
ptr = other.ptr;
other.ptr = nullptr;
```

Reply:
I've seen this done as a swap somewhere, but nulling other is fine too. Just trying to be "fancy" and I'm not too attached to it.

I can do it the other way.

Reply:
`std::exchange` would probably be the correct "fancy".  Please don't try to be fancy.

Reply:
It was a specific thing I remember from an undergrad C++ course actually, something about re-using constructor/destructor to implement assignment. Yes, specifically with `swap`... but I either misremembered it, or it's not useful here.

Not so much "trying" to be fancy, as ending up doing it without realizing occasionally.